hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/utils/TokenInfoRetrievalUtils.java

```diff
@@ -0,0 +1,241 @@
+import static com.hedera.services.store.contracts.precompile.TokenKeyType.PAUSE_KEY;
+import static com.hedera.services.store.contracts.precompile.TokenKeyType.SUPPLY_KEY;
+import static com.hedera.services.store.contracts.precompile.TokenKeyType.WIPE_KEY;
+
+import com.hedera.services.config.NetworkInfo;
+import com.hedera.services.legacy.core.jproto.JKey;
+import com.hedera.services.state.enums.TokenSupplyType;
+import com.hedera.services.state.submerkle.FcCustomFee;
+import com.hedera.services.store.contracts.WorldLedgers;
+import com.hedera.services.store.contracts.precompile.TokenKeyType;
+import com.hedera.services.store.contracts.precompile.codec.Expiry;
+import com.hedera.services.store.contracts.precompile.codec.FixedFee;
+import com.hedera.services.store.contracts.precompile.codec.FractionalFee;
+import com.hedera.services.store.contracts.precompile.codec.FungibleTokenInfo;
+import com.hedera.services.store.contracts.precompile.codec.HederaToken;
+import com.hedera.services.store.contracts.precompile.codec.KeyValue;
+import com.hedera.services.store.contracts.precompile.codec.NonFungibleTokenInfo;
+import com.hedera.services.store.contracts.precompile.codec.RoyaltyFee;
+import com.hedera.services.store.contracts.precompile.codec.TokenInfo;
+import com.hedera.services.store.contracts.precompile.codec.TokenKey;
+import com.hedera.services.store.models.NftId;
+import com.hedera.services.utils.EntityIdUtils;
+import com.hederahashgraph.api.proto.java.TokenID;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.datatypes.Address;
+
+public class TokenInfoRetrievalUtils {
+
+    private TokenInfoRetrievalUtils() {
+        throw new UnsupportedOperationException("Utility class");
+    }
+
+    public static NonFungibleTokenInfo getNonFungibleTokenInfo(
+            final TokenID tokenId,
+            final long serialNumber,
+            final WorldLedgers ledgers,
+            final NetworkInfo networkInfo) {
+        final var tokenInfo = getTokenInfo(tokenId, ledgers, networkInfo);
+        final var nftId = NftId.fromGrpc(tokenId, serialNumber);
+
+        final var ownerId = ledgers.ownerOf(nftId);
+        final var creationTime = ledgers.packedCreationTimeOf(nftId);
+        final var metadata = ledgers.metadataOf(nftId);
+        final var spenderId = EntityIdUtils.asTypedEvmAddress(ledgers.spenderOf(nftId));
+        return new NonFungibleTokenInfo(
+                tokenInfo, serialNumber, ownerId, creationTime, metadata, spenderId);
+    }
+
+    public static FungibleTokenInfo getFungibleTokenInfo(
+            final TokenID tokenId, final WorldLedgers ledgers, final NetworkInfo networkInfo) {
+        final var tokenInfo = getTokenInfo(tokenId, ledgers, networkInfo);
+        final var decimals = ledgers.decimalsOf(tokenId);
+        return new FungibleTokenInfo(tokenInfo, decimals);
+    }
+
+    public static TokenInfo getTokenInfo(
```

Comment:
Having each field go through `ledgers.<foo>(tokenId)` is churning a lot of lookups that should be constrained to one.  Consider a singular `ledgers.tokenInfo(tokenId)` method and remove all the methods that are used just one time to fill just one field.